### ISSUE TITLE
Web (EmptyWeb) F# version

### ISSUE THREAD

**Issue by peterblazejewicz** [dotnet/templating#251]
Should F# version of `web` project (Web/Empty) be added?
We provide one with `generator-aspnet` and I'm rewriting it for RC:
https://github.com/OmniSharp/generator-aspnet/tree/master/templates/projects/fsharp_emptyweb
https://github.com/OmniSharp/generator-aspnet/issues/881
Maybe not now, but do you want a PR?

Thanks!

**Comment by mlorbetske** [comment]
A PR would be great, thanks!

**Comment by mlorbetske** [comment]
@seancpeters can you help out with anything @peterblazejewicz needs with these (this one and #252)?

**Comment by seancpeters** [comment]
Definitely. Please let me know what I can do to help.

**Comment by peterblazejewicz** [comment]
@seancpeters 
I'll add a link to a PR with changes in other repo. You could just get it and make your inputs (e.g. changes according to source code specs/requirements/etc)

**Comment by peterblazejewicz** [comment]
@seancpeters 
I you don't mind review: 
https://github.com/OmniSharp/generator-aspnet/pull/896
Project files:
https://github.com/peterblazejewicz/generator-aspnet/tree/2967c35294d32a985c2792e4f21499b27010a14d/templates/projects/fsharp_emptyweb

**Comment by seancpeters** [comment]
To make sure I understand correctly: I can take the template from those links, and make one for dotnet/templating that has identical output?

**Comment by peterblazejewicz** [comment]
> To make sure I understand correctly: I can take the template from those links, and make one for dotnet/templating that has identical output?

It would be enough to just add comment(s) on that PR. I could do PR here using the same, adapted content (at least try to do one, don't know a project details yet)

**Comment by seancpeters** [comment]
It's probably easiest if I discuss here what needs to be done.

Let's use the C# empty web template as an example:
https://github.com/dotnet/templating/tree/master/template_feed/Microsoft.DotNet.Web.ProjectTemplates/content/EmptyWeb-CSharp

For an F# empty web template, you'll want to make a directory /EmptyWeb-FSharp/ at the same level as the /EmptyWeb-CSharp/ directory. Everything for the F# template will go under /EmptyWeb-FSharp/
It's probably easiest to name the files similar to how they're named in the C# template - you'll want:
- Company.WebApplication1.fsproj
- Program.fs
- Startup.fs

We also need to look at the template.json file for the C# emptyweb project:
https://github.com/dotnet/templating/blob/master/template_feed/Microsoft.DotNet.Web.ProjectTemplates/content/EmptyWeb-CSharp/.template.config/template.json

You'll need to make a /EmptyWeb-FSharp/.template.config/template.json file for the F# emptyweb project. You can copy the C# one, and just make a few changes. Here's the start of the file, it's the only place changes will need to be made:
```

{
  "author": "Microsoft",
  "classifications": ["Web", "Empty"],
  "name": "Empty ASP.NET Core Web Application",
  "groupIdentity": "Microsoft.Web.Empty",
  "identity": "Microsoft.Web.Empty.CSharp",
  "shortName": "web",
  "tags": {
    "language": "C#",
    "type": "project"
  },
  "sourceName": "Company.WebApplication1",
  "preferNameDirectory": true,
  "symbols": {
    "IncludeApplicationInsights": {
      "type": "parameter",
      "datatype": "bool",
      "defaultValue": "false",
      "description": "Whether or not to include Application Insights in the project"
    },
    "Framework": {
      "type": "parameter",
      "datatype": "choice",
      "choices": [
        {
          "choice": "netcoreapp1.0",
          "description": "Target netcoreapp1.0"
        },
        {
          "choice": "netcoreapp1.1",
          "description": "Target netcoreapp1.1"
        }
      ],
      "defaultValue": "netcoreapp1.0"
    },
```

I don't think your F# template has Application Insights built in, so we can remove the IncludeApplicationInsights section under the "symbols".
You'll want to change "language": "C#"
... to be "language": "F#"
And change the   "identity": "Microsoft.Web.Empty.CSharp"
... to be   "identity": "Microsoft.Web.Empty.FSharp",
No other changes should be necessary in the template.json file (copy)


This part doens't need to be changed. But it is important for the changes we'll need to make in the .fs files:   "sourceName": "Company.WebApplication1",
When you create a template, the engine determines a string value for "sourceName". 

Sidenote: How the sourceName is determined with the command "dotnet new":
	- if you provide "-n Name"option, then sourceName = Name
	- if you do not provide "-n Name", but provide a "-o DirectoryName" option, then the sourceName = DirectoryName
	- if you specify both "-o DirectoryName" and "-n Name", then sourceName = Name
	- if you don't specify either option, then sourceName = the name of the current directory

That value will replace the string literal "Company.WebApplication1" in all the filenames, and in all the file content in the template.
For example, if I create a template with: 
dotnet new web -n MyWebProject
all instances of "Company.WebApplication1" will be replaced with "MyWebProject".

Since we are naming the proj file "Company.WebApplication1.fsproj" in the template, the output file name will become "MyWebProject.fsproj"

The same idea should be used for the namespace value in the .fs files. In the Program.fs & Startup.fs files, you have the line:
open <%= namespace %>

For the dotnet template, this will become:
open Company.WebApplication1

... and when the template is generated, the output .fs files will contain the line:
open MyWebProject


The .fsproj file contains this line:
`  <TargetFramework><%= dotnet.targetFramework %></TargetFramework>`
For dotnet templates, we do this differently. Under the "symbols" section of the template.json file, there is a "Framework" section. Since it has "type": "parameter", it's something the user can specify.
We use the specified value to format the TargetFramework. You can copy exactly what is in the existing C# template, specifically:
```
  <PropertyGroup>
    <TargetFramework Condition="'$(Framework)' != 'netcoreapp1.1'">netcoreapp1.0</TargetFramework>
    <TargetFramework Condition="'$(Framework)' == 'netcoreapp1.1'">netcoreapp1.1</TargetFramework>
```

The template engine evaluates the Condition statements, the ones that are true are written to the target, others are not.

There is one more extra file to add to your template. It should be:
/EmptyWeb-FSharp/.template.config/dotnetcli.host.json
and can contain exactly this:
```
{
  "symbolInfo": {
    "Framework": {
      "longName": "framework"
    }
  }
}
```

Everything else in the template is general-purpose. This file will only be used by the command line "dotnet new". It's just saying that for the parameter "Framework", we want the parameter to be specified with "framework" instead.

I think I covered everything your template already does. If I missed anything, or if anything is unclear, please let me know.

**Comment by peterblazejewicz** [comment]
> I don't think your F# template has Application Insights built in, so we can remove the IncludeApplicationInsights section under the "symbols".

Folks, what is the UX pattern to get this option into `dotnet new`? By default it has `false` flag, so never gets into output. I could't find a reference in order to decide if Application Insight should be used in templating in Yeoman stuff:
https://github.com/dotnet/templating/search?p=2&q=IncludeApplicationInsights

**Comment by mlorbetske** [comment]
If you want it to be on by default, you can switch the default to true in template.json & remove marking the parameter as hidden in dotnetcli.host.json so that users can see the switch. Someone could technically turn it on at the CLI already if they pass --UseApplicationInsights (I believe is the name of the parameter)

**Comment by lambdakris** [comment]
Is this still up for grabs? If so, I would like to help with this. Looking at the current state of the dev branch, my only immediate question would be whether the template should be for/under the 1.x or 2.0 directories...or both?

**Comment by mlorbetske** [comment]
Yep, it's still up for grabs, we'd be happy to have the help :) Ideally, we'd add a 1.0 & 1.1 flavor of the template in the 1.x directory and a 2.0 in the 2.0 directory - we're separating the delivery of the 1.x templates from the 2.0 (and future ones) so that there's flexibility in which templates accompany the runtimes installed by the CLI/VS/VS for mac. If you don't want to do both though, we can do the port from 2.0 back to 1.x.

**Comment by lambdakris** [comment]
What's the correct approach to testing/verification of a built in template? Is installing the template with dotnet new --install sufficient?

**Comment by mlorbetske** [comment]
That's an option, the other way (what we do when we're getting ready to release a new CLI) is to treat the template with new3 (built with setup.cmd or setup.sh) since that engine will ship at the same time.

**Comment by lambdakris** [comment]
I was able to manually test with new3 and created a pr https://github.com/dotnet/templating/pull/600 for the 1.x version. I'll wait on feedback then work on the 2.0 version.

**Comment by lambdakris** [comment]
So now I am looking to work on the 2.0 version of the template. For reference, I compared the 1.x and 2.0 versions of the EmptyWeb-CSharp template to one another and could only find the following diffs:
1. Targeting netcoreapp2.0
2. Referencing AspNetCore 2.0.0-preview1-*
3. Presence of NuGet.config with package sources to myget feeds where presumably the wip 2.0 packages are hosted
4. Logging config moved from Startup class to Program class
5. No App Insights

Is there anything else I should take note off for the 2.0 version of this template?

**Comment by lambdakris** [comment]
Also, getting this error during `dotnet restore` in the EmptyWeb-CSharp template:
```
...
C:\Program Files\dotnet\sdk\1.0.3\NuGet.targets(97,5): error : Unable to resolve 'Microsoft.NETCore.App (>= 2.0.0)' for '.NETCoreApp,Version=v2.0'. [C:\Users\lambdakris\Code\templating\template_feed\Microsoft.DotNet.Web.ProjectTemplates.2.0\content\EmptyWeb-CSharp\Company.WebApplication1.csproj]
...
```

Is this because I need to have some kind of nightly version of .NET Core 2.0 runtime or SDK installed or is it a problem retrieving something from the myget feeds?

**Comment by mlorbetske** [comment]
Yep, it needs the nightly of the 2.0 cli unfortunately. I believe that's the total set of differences, @glennc can probably point out any changes that are coming.
